File: hw/saraguro/u1/HW08 SaveAnObject/BookSaraguroJS/src/controller/LibraryManager.js
const readlineSync = require('readline-sync');
const FileManager = require('./fileManager');
const Book = require('./book');

class LibraryManager {
    constructor(filename) {
        this.filename = filename;
        this.books = FileManager.loadBooksFromFile(filename);
    }

    run() {
        let option;

        do {
            console.log("\nWelcome to the Library System");
            console.log("1. Add Book");
            console.log("2. Display All Books");
            console.log("3. Delete Book");
            console.log("4. Save and Exit");
            option = readlineSync.question("Choose an option: ");

            switch (option) {
                case '1':
                    this.addBook();
                    break;
                case '2':
                    this.displayAllBooks();
                    break;
                case '3':
                    this.deleteBook();
                    break;
                case '4':
                    FileManager.saveBooksToFile(this.filename, this.books);
                    console.log("Books saved. Exiting...");
                    break;
                default:
                    console.log("Invalid option. Please try again.");
            }
        } while (option !== '4');
    }

    addBook() {
        const title = readlineSync.question("Enter title: ");
        const author = readlineSync.question("Enter author: ");
        const year = readlineSync.questionInt("Enter year: ");
        const publisher = readlineSync.question("Enter publisher: ");
        const genre = readlineSync.question("Enter genre: ");

        const book = new Book(title, author, year, publisher, genre);
        this.books.push(book);
    }

    displayAllBooks() {
        if (this.books.length === 0) {
            console.log("No books to display.");
            return;
        }
        this.books.forEach((book, index) => {
            console.log(`${index + 1}. ${book.toString()}`);
        });
    }

    deleteBook() {
        this.displayAllBooks();
        if (this.books.length === 0) {
            return;
        }

        const index = readlineSync.questionInt("Enter the number of the book to delete: ") - 1;

        if (index < 0 || index >= this.books.length) {
            console.log("Invalid book number.");
            return;
        }

        const removedBook = this.books.splice(index, 1);
        console.log(`Book removed: ${removedBook[0].title}`);
    }
}

module.exports = LibraryManager;

File: hw/saraguro/u1/HW08 SaveAnObject/BookSaraguro/src/ec/edu/espe/booksaraguro/controller/LibraryManager.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package ec.edu.espe.booksaraguro.controller;

import ec.edu.espe.booksaraguro.model.Book;
import java.util.ArrayList;
import java.util.Scanner;
import utils.BookFileManager;

/**
 *
 * @author Leidy Saraguro,Paradigm Pioneers Squad, DCCO-ESPE
 */
public class LibraryManager {
    private ArrayList<Book> books;
    private final BookFileManager fileManager;
    private String filename;

    public LibraryManager() {
        fileManager = new BookFileManager();
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter filename for books data (e.g., books.dat): ");
        filename = scanner.nextLine();
        books = fileManager.loadBooksFromFile(filename);
    }

    public void run() {
        Scanner scanner = new Scanner(System.in);
        int option;

        do {
            System.out.println("Welcome to the Library System");
            System.out.println("1. Add Book");
            System.out.println("2. Display All Books");
            System.out.println("3. Delete Book");
            System.out.println("4. Save and Exit");
            System.out.print("Choose an option: ");
            option = scanner.nextInt();
            scanner.nextLine();  // Consume newline

            switch (option) {
                case 1:
                    addBook(scanner);
                    break;
                case 2:
                    displayAllBooks();
                    break;
                case 3:
                    deleteBook(scanner);
                    break;
                case 4:
                    fileManager.saveBooksToFile(filename, books);
                    System.out.println("Books saved. Exiting...");
                    break;
                default:
                    System.out.println("Invalid option. Please try again.");
            }
        } while (option != 4);

        scanner.close();
    }

    private void addBook(Scanner scanner) {
        System.out.print("Enter title: ");
        String title = scanner.nextLine();
        System.out.print("Enter author: ");
        String author = scanner.nextLine();
        System.out.print("Enter year: ");
        int year = scanner.nextInt();
        scanner.nextLine();  // Consume newline
        System.out.print("Enter publisher: ");
        String publisher = scanner.nextLine();
        System.out.print("Enter genre: ");
        String genre = scanner.nextLine();

        Book book = new Book(title, author, year, publisher, genre);
        books.add(book);
    }

    private void displayAllBooks() {
        for (int i = 0; i < books.size(); i++) {
            System.out.println((i + 1) + ". " + books.get(i));
        }
    }

    private void deleteBook(Scanner scanner) {
        displayAllBooks();
        if (books.isEmpty()) {
            System.out.println("No books to delete.");
            return;
        }

        System.out.print("Enter the number of the book to delete: ");
        int index = scanner.nextInt();
        scanner.nextLine();  // Consume newline

        if (index < 1 || index > books.size()) {
            System.out.println("Invalid book number.");
            return;
        }

        Book removedBook = books.remove(index - 1);
        System.out.println("Book removed: " + removedBook.getTitle());
    }
}
